File: test-automation-core/src/main/java/com/nortal/test/core/services/report/cucumber/RequestResponseReportFormatter.java
package com.nortal.test.core.services.report.cucumber;

import static com.nortal.test.core.services.report.cucumber.JsonFormattingUtils.prettyPrintHtmlJson;
import static com.nortal.test.core.services.report.cucumber.JsonFormattingUtils.prettyPrintJson;

import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.nortal.test.core.rest.RetrofitCallHelper;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import org.apache.commons.lang3.StringUtils;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class RequestResponseReportFormatter {
	private static final Gson GSON = new GsonBuilder()
			.serializeNulls()
			.setPrettyPrinting()
			.create();

	private final ReportFormatter formatter;

	/**
	 * Formats the request and response and adds it to the report.
	 *
	 * @param requestSpec request
	 * @param response    response
	 */
	public void formatAndAddToReport(final Request requestSpec, final Response response) {
		final ReportFormatter.Attachment attachment = populateRequestSection(ReportFormatter.Attachment.create(), requestSpec)
				.addSection("Response:", ReportFormatter.SectionType.COLLAPSIBLE, getResponse(response))
				/*.addSection("Registered header overrides:", ReportFormatter.SectionType.COLLAPSIBLE, getHeaderOverrides())*/;
		formatter.formatAndAddToReport(attachment);
	}

	public void formatAndAddToReport(final Request requestSpec) {
		final ReportFormatter.Attachment attachment = populateRequestSection(ReportFormatter.Attachment.create(), requestSpec);
		formatter.formatAndAddToReport(attachment);
	}

	private ReportFormatter.Attachment populateRequestSection(final ReportFormatter.Attachment attachment,
	                                                          final Request requestSpec) {
		attachment
				.setName(requestSpec.method().toUpperCase(Locale.ENGLISH) + " " + requestSpec.url().toString())
				.setTitle(getRequestDescription(requestSpec))
				.addSection("Request:", ReportFormatter.SectionType.STANDARD, getRequest(requestSpec));

		return attachment;
	}

	private String getRequest(final Request request) {
		final StringBuilder curl = new StringBuilder("curl -X ")
				.append(request.method().toUpperCase(Locale.ENGLISH))
				.append(" \\\n'")
				.append(request.url().toString())
				.append("' \\\n");

		request.headers()
				.toMultimap()
				.forEach((headerKey, values) -> values.forEach(headerValue -> appendHeader(curl, headerKey, headerValue)));

		if (request.body() != null) {

			final MediaType mediaType = request.body().contentType();
			if (mediaType != null) {
				appendHeader(curl, "content-type", mediaType.toString());
			}

			final String json = prettyPrintJson(RetrofitCallHelper.convertRequestBodyToString(request));
			curl.append("-d '").append(json).append('\'');
		}
		curl.append('\n');

		return curl.toString();
	}

	private void appendHeader(final StringBuilder curl, final String headerKey, final String headerValue) {
		curl.append("-H '").append(headerKey).append(": ").append(headerValue).append("' \\\n");
	}

	@SneakyThrows
	private String getResponse(final Response response) {
		final Map<String, Object> responseMap = new HashMap<>();
		responseMap.put("status", response.code());
		responseMap.put("headers", response.headers().toMultimap());

		final ResponseBody body = response.peekBody(Long.MAX_VALUE);
		final String bodyStr = body.string();
		if (StringUtils.isNotBlank(bodyStr)) {
			try {
				final JsonElement element = GSON.fromJson(bodyStr, JsonElement.class);
				responseMap.put("body", element);
			} catch (JsonParseException ex) {
				responseMap.put("body", bodyStr);
			}
		}

		return prettyPrintHtmlJson(responseMap);
	}

	private String getRequestDescription(final Request requestSpec) {
		return String.format("Rest call: %s", requestSpec.url().toString());
	}

	/*private String getHeaderOverrides() {
		final EnvConfigurationOverride override = headerOverrideService.getHeaderConfigurationOverrides();
		return prettyPrintHtmlJson(override);
	}*/
}

File: test-automation-core/src/main/kotlin/com/nortal/test/core/report/RequestResponseReportFormatter.kt
/**
 * Copyright (c) 2022 Nortal AS
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.nortal.test.core.report

import com.fasterxml.jackson.core.JacksonException
import com.fasterxml.jackson.databind.ObjectMapper
import com.nortal.test.core.report.model.HttpRequest
import com.nortal.test.core.report.model.HttpResponse
import com.nortal.test.core.report.util.CurlBuilder
import com.nortal.test.core.report.util.HttpContentTypeUtils
import org.apache.commons.lang3.StringUtils
import org.springframework.http.HttpHeaders
import org.springframework.stereotype.Component

/**
 * TODO: class  requires improvements
 */
@Component
open class RequestResponseReportFormatter(
    private val objectMapper: ObjectMapper,
    private val curlBuilder: CurlBuilder,
    private val formatter: ReportFormatter
) {

    /**
     * Formats the request and response and adds it to the report.
     *
     * @param request request
     * @param response    response
     */
    fun formatAndAddToReport(request: HttpRequest, response: HttpResponse) {
        val attachment = populateRequestSection(ReportFormatter.Attachment.create(), request)
            .addSection(
                "Response:",
                ReportFormatter.SectionType.COLLAPSIBLE,
                getResponse(response)
            )

        addAdditionalSections(request, response, attachment)
        formatter.formatAndAddToReport(attachment)
    }

    open fun addAdditionalSections(
        request: HttpRequest,
        response: HttpResponse,
        attachment: ReportFormatter.Attachment
    ) {
        //do nothing by default.
    }

    fun formatAndAddToReport(requestSpec: HttpRequest) {
        val attachment = populateRequestSection(ReportFormatter.Attachment.create(), requestSpec)
        formatter.formatAndAddToReport(attachment)
    }

    private fun populateRequestSection(
        attachment: ReportFormatter.Attachment,
        requestSpec: HttpRequest
    ): ReportFormatter.Attachment {
        attachment
            .setName(requestSpec.method.uppercase() + " " + requestSpec.url)
            .setTitle(getRequestDescription(requestSpec))
            .addSection("Request:", ReportFormatter.SectionType.STANDARD, curlBuilder.getRequestCurl(requestSpec))
        return attachment
    }


    private fun getResponse(response: HttpResponse): String {
        val responseMap: MutableMap<String, Any> = HashMap()
        responseMap["status"] = response.statusCode
        responseMap["headers"] = response.headers

        val bodyStr = response.body
        val contentType = getContentType(response)
        if (StringUtils.isNotBlank(bodyStr)) {
            if (isJsonContentType(contentType)) {
                responseMap["body"] = resolveJsonBody(bodyStr!!)
            } else if (HttpContentTypeUtils.isXmlContentType(contentType)) {
                responseMap["body"] = resolveXmlBody(bodyStr!!)
            } else {
                responseMap["body"] = "Endpoint returned unknown content type, not putting it to the report"
            }
        }
        return JsonFormattingUtils.prettyPrintHtml(responseMap)
    }

    protected open fun isJsonContentType(contentType: String?): Boolean {
        return HttpContentTypeUtils.isJsonContentType(contentType)
    }

    protected open fun resolveJsonBody(body: String): Any {
        return try {
            objectMapper.readTree(body)
        } catch (ex: JacksonException) {
            body
        }
    }

    protected open fun resolveXmlBody(body: String): Any {
        //TODO for now returning as is.
        return body
    }

    private fun getContentType(response: HttpResponse): String? {
        return response.headers.getFirstByName(HttpHeaders.CONTENT_TYPE)
    }


    private fun getRequestDescription(requestSpec: HttpRequest): String {
        return String.format("Rest call: %s", requestSpec.url)
    }


}